shell/platform/linux/fl_json_message_codec.cc

```diff
@@ -0,0 +1,300 @@
+          return FALSE;
+      writer.EndArray();
+      break;
+    case FL_VALUE_TYPE_MAP:
+      writer.StartObject();
+      for (size_t i = 0; i < fl_value_get_length(value); i++) {
+        FlValue* key = fl_value_get_map_key(value, i);
+        if (fl_value_get_type(key) != FL_VALUE_TYPE_STRING) {
+          g_set_error(error, FL_JSON_MESSAGE_CODEC_ERROR,
+                      FL_JSON_MESSAGE_CODEC_ERROR_INVALID_OBJECT_KEY_TYPE,
+                      "Invalid object key type");
+          return FALSE;
+        }
+        writer.Key(fl_value_get_string(key));
+        if (!write_value(writer, fl_value_get_map_value(value, i), error))
+          return FALSE;
+      }
+      writer.EndObject();
+      break;
+    default:
+      g_set_error(error, FL_MESSAGE_CODEC_ERROR, FL_MESSAGE_CODEC_ERROR_FAILED,
+                  "Encountered unknown FlValue type");
+      return FALSE;
+  }
+
+  return TRUE;
+}
+
+// Handler to parse JSON using rapidjson in SAX mode
+struct FlValueHandler {
+  GPtrArray* stack;
+  FlValue* key;
+
+  FlValueHandler() {
+    stack = g_ptr_array_new_with_free_func(
+        reinterpret_cast<GDestroyNotify>(fl_value_unref));
+    key = nullptr;
+  }
+
+  ~FlValueHandler() {
+    g_ptr_array_unref(stack);
+    if (key != nullptr)
+      fl_value_unref(key);
+  }
+
+  // Get the current head of the stack
+  FlValue* get_head() {
+    if (stack->len == 0)
+      return nullptr;
+    return static_cast<FlValue*>(g_ptr_array_index(stack, stack->len - 1));
+  }
+
+  // Push a value onto the stack
+  void push(FlValue* value) { g_ptr_array_add(stack, fl_value_ref(value)); }
+
+  // Pop the stack
+  void pop() { g_ptr_array_remove_index(stack, stack->len - 1); }
+
+  // Add a new value to the stack
+  bool add(FlValue* value) {
+    g_autoptr(FlValue) owned_value = value;
+    FlValue* head = get_head();
+    if (head == nullptr)
+      push(owned_value);
+    else if (fl_value_get_type(head) == FL_VALUE_TYPE_LIST)
+      fl_value_append(head, owned_value);
+    else if (fl_value_get_type(head) == FL_VALUE_TYPE_MAP) {
+      fl_value_set_take(head, key, fl_value_ref(owned_value));
+      key = nullptr;
+    } else
+      return false;
+
+    if (fl_value_get_type(owned_value) == FL_VALUE_TYPE_LIST ||
+        fl_value_get_type(owned_value) == FL_VALUE_TYPE_MAP)
+      push(value);
+
+    return true;
+  }
+
+  // The following implements the rapidjson SAX API
+
+  bool Null() { return add(fl_value_new_null()); }
+
+  bool Bool(bool b) { return add(fl_value_new_bool(b)); }
+
+  bool Int(int i) { return add(fl_value_new_int(i)); }
+
+  bool Uint(unsigned i) { return add(fl_value_new_int(i)); }
+
+  bool Int64(int64_t i) { return add(fl_value_new_int(i)); }
+
+  bool Uint64(uint64_t i) {
+    return add(fl_value_new_int(i));  // FIXME: Too big!
+  }
+
+  bool Double(double d) { return add(fl_value_new_float(d)); }
+
+  bool RawNumber(const char* str, rapidjson::SizeType length, bool copy) {
+    return false;
```

Comment:
Does this need any error handling, or is it impossible to reach? (If the latter, maybe an assertion?)

Reply:
I *believe* returning false here causes rapidjson to fail the decoding, which means we'll generate a `GError` in this case. So it's possible to reach but we'll do the right thing if it is.

Reply:
I changed it to generate an error if this is hit. I couldn't find a case when it is called.